src/reference_update.rs

```diff
@@ -0,0 +1,146 @@
+use crate::error::CapnError;
+use git2::Oid;
+use std::error::Error;
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum ReferenceUpdate {
+    New {
+        new_commit_id: Oid,
+        ref_name: String,
+    },
+    Delete {
+        old_commit_id: Oid,
+        ref_name: String,
+    },
+    Update {
+        old_commit_id: Oid,
+        new_commit_id: Oid,
+        ref_name: String,
+    },
+}
+
+impl ReferenceUpdate {
+    pub fn from_git_hook_format(
+        old_commit_id: &str,
+        new_commit_id: &str,
+        ref_name: &str,
+    ) -> Result<ReferenceUpdate, Box<dyn Error>> {
+        let old_commit_id = Oid::from_str(old_commit_id)?;
+        let new_commit_id = Oid::from_str(new_commit_id)?;
+        let ref_name = ref_name.to_owned();
+        match (old_commit_id.is_zero(), new_commit_id.is_zero()) {
+            (false, false) => Ok(ReferenceUpdate::Update {
+                old_commit_id,
+                new_commit_id,
+                ref_name,
+            }),
+            (false, true) => Ok(ReferenceUpdate::Delete {
+                old_commit_id,
+                ref_name,
+            }),
+            (true, false) => Ok(ReferenceUpdate::New {
+                new_commit_id,
+                ref_name,
+            }),
+            (true, true) => Err(Box::new(CapnError::new("Invalid reference update specification, trying to update from a zero commit to another zero commit")))
+        }
+    }
+
+    pub fn old_commit_id(&self) -> Option<Oid> {
```

Comment:
Not sure how I feel about these functions.
They bleed through details about the concrete type which would have been available via pattern matching anyway.
Just something to think about...

Reply:
Hmm. I agree that it feels like it's leaking details, but I'm not sure if it necessarily is. I think there are two contexts that this ADT is used.
- In some of them, your business logic cares that this was a delete, or that you're updating something that existed. For example, new branches don't follow all the same rules. These cases definitely should be pattern matching the cases they care about.
- In others, you only care to get some previous point if one exists. For example, getting the range of commits affected. These cases shouldn't care about the individual cases that could result in the commit you're coming from existing.

I think we can both agree that using an Option here is better than just treating 00000 is a null-commit.

Reply:
Idd.  Definitely agree that it's a big improvement on the previous situation.
It seems to indicate to me that there's something else going on if this needs to be viewed differently from different perspectives, but since all I have is a vague feeling, I'm going to go ahead and merge this if you're happy that it's ready?